packages/components/src/modal/index.tsx

```diff
@@ -39,6 +39,34 @@ import type { ModalProps } from './types';
 // Used to count the number of open modals.
 let openModalCount = 0;
 
+/**
+ * When `firstElement` is passed to `focusOnMount`, this function is optimized to
+ * avoid focusing on the `Close` button (or other "header" elements of the Modal
+ * and instead focus within the Modal's contents.
+ * However, if no tabbable elements are found within the Modal's contents, the
+ * first tabbable element (likely the `Close` button) will be focused instead.
+ * This ensures that at least one element is focused whilst still optimizing
+ * for the best a11y experience.
+ *
+ * See: https://github.com/WordPress/gutenberg/issues/54106.
+ * @param tabbables Element[] an array of tabbable elements.
+ * @return Element the first tabbable element in the Modal contents (or any tabbable element if none are found in content).
+ */
+function getFirstTabbableElement( tabbables: Element[] ) {
+	// Attempt to locate tabbable outside of the header portion of the Modal.
+	const firstContentTabbable = tabbables.find( ( tabbable ) => {
+		return tabbable.closest( '.components-modal__header' ) === null;
+	} );
+
+	if ( firstContentTabbable ) {
+		return firstContentTabbable;
+	}
+
+	// Fallback to the first tabbable element anywhere within the Modal.
+	// Likely the `Close` button.
+	return tabbables[ 0 ];
```

Comment:
Nit: this could be simplified to

```suggestion
	return (
		// Attempt to locate tabbable outside of the header portion of the Modal.
		tabbables.find(
			( tabbable ) =>
				tabbable.closest( `.${ MODAL_HEADER_CLASSNAME }` ) === null
		) ??
		// Fallback to the first tabbable element anywhere within the Modal.
		// Likely the `Close` button.
		tabbables[ 0 ]
	);
```